src/bot.ts

```diff
@@ -525,33 +537,79 @@ you can circumvent this protection against memory leaks.`);
 /**
  * Performs a network call task, retrying upon known errors until success.
  *
+ * If the task errors and a retry_after value can be used, a subsequent retry
+ * will be delayed by the specified period of time.
+ *
+ * Otherwise, if the first attempt at running the task fails, the task is
+ * retried immediately. If second attempt fails, too, waits for 100 ms, and then
+ * doubles this delay for every subsequent attemt. Never waits longer than 1
+ * hour before retrying.
+ *
  * @param task Async task to perform
+ * @param signal Optional `AbortSignal` to prevent further retries
  */
-async function withRetries<T>(task: () => Promise<T>): Promise<T> {
+async function withRetries<T>(
+    task: () => Promise<T>,
+    signal?: AbortSignal,
+): Promise<T> {
     let result: { ok: false } | { ok: true; value: T } = { ok: false };
+    const INITIAL_DELAY = 100; // ms
+    let delay = INITIAL_DELAY;
     while (!result.ok) {
+        let mustDelay = true;
         try {
             result = { ok: true, value: await task() };
+            mustDelay = false;
         } catch (error) {
             debugErr(error);
             if (error instanceof HttpError) continue;
             if (error instanceof GrammyError) {
                 if (error.error_code >= 500) continue;
                 if (error.error_code === 429) {
                     const retryAfter = error.parameters.retry_after;
-                    if (retryAfter !== undefined) await sleep(retryAfter);
+                    if (retryAfter !== undefined) {
+                        await sleep(retryAfter, signal);
+                        mustDelay = false;
+                    }
                     continue;
                 }
             }
             throw error;
+        } finally {
```

Comment:
This should not be entered when the error is rethrown. This needs to be reworked before it can be merged.

Reply:
Fixed by #379.